File: mobile/modules/ikkyee-place-search/ios/IkkyeePlaceSearchModule.swift
import CoreLocation
import ExpoModulesCore
import GooglePlaces

public final class IkkyeePlaceSearchModule: Module {
  private var initialized = false

  public func definition() -> ModuleDefinition {
    Name("IkkyeePlaceSearch")

    AsyncFunction("searchPlaces") { (
      query: String,
      north: Double,
      south: Double,
      east: Double,
      west: Double,
      promise: Promise
    ) in
      guard validInput(query, north, south, east, west),
            Bundle.main.bundleIdentifier == expectedBundleId else {
        promise.reject("E_PLACE_SEARCH_INVALID", "Native place search input is invalid")
        return
      }
      guard let apiKey = Bundle.main.object(forInfoDictionaryKey: "GMSApiKey") as? String,
            !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        promise.reject("E_PLACE_SEARCH_UNAVAILABLE", "Native place search is unavailable")
        return
      }
      if !initialized {
        GMSPlacesClient.provideAPIKey(apiKey)
        initialized = true
      }

      let properties = [
        GMSPlaceProperty.placeID,
        GMSPlaceProperty.name,
        GMSPlaceProperty.formattedAddress,
        GMSPlaceProperty.coordinate
      ].map { $0.rawValue }
      let request = GMSPlaceSearchByTextRequest(
        textQuery: query.trimmingCharacters(in: .whitespacesAndNewlines),
        placeProperties: properties
      )
      request.maxResultCount = 5
      request.locationBias = GMSPlaceRectangularLocationOption(
        northEast: CLLocationCoordinate2D(latitude: north, longitude: east),
        southWest: CLLocationCoordinate2D(latitude: south, longitude: west)
      )

      GMSPlacesClient.shared().searchByText(with: request) { places, error in
        if let error {
          rejectPlaceSearchError(error, promise)
          return
        }
        guard let places else {
          promise.reject("E_PLACE_SEARCH", "Native place search failed")
          return
        }
        let projection: [[String: Any]] = places.prefix(5).compactMap { place in
          guard let id = place.placeID, !id.isEmpty,
                let name = place.name, !name.isEmpty else { return nil }
          return [
            "id": id,
            "name": name,
            "address": place.formattedAddress?.isEmpty == false ? place.formattedAddress! : name,
            "latitude": place.coordinate.latitude,
            "longitude": place.coordinate.longitude
          ]
        }
        promise.resolve(projection)
      }
    }
  }
}

private func rejectPlaceSearchError(_ error: Error, _ promise: Promise) {
  let providerError = error as NSError
  guard providerError.domain == kGMSPlacesErrorDomain,
        let placesCode = GMSPlacesErrorCode(rawValue: providerError.code) else {
    promise.reject("E_PLACE_SEARCH", "Native place search failed")
    return
  }
  switch placesCode {
  case .networkError:
    promise.reject("E_PLACE_SEARCH_NETWORK", "Native place search network failed")
  case .usageLimitExceeded, .rateLimitExceeded, .deviceRateLimitExceeded:
    promise.reject("E_PLACE_SEARCH_QUOTA", "Native place search quota exceeded")
  case .keyInvalid, .keyExpired, .accessNotConfigured, .incorrectBundleIdentifier:
    promise.reject("E_PLACE_SEARCH_CONFIGURATION", "Native place search configuration rejected")
  default:
    promise.reject("E_PLACE_SEARCH", "Native place search failed")
  }
}

private let expectedBundleId = "com.ikkyee.mobile"

private func validInput(_ query: String, _ north: Double, _ south: Double, _ east: Double, _ west: Double) -> Bool {
  let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
  return !trimmed.isEmpty && trimmed.count <= 80 &&
    north.isFinite && south.isFinite && east.isFinite && west.isFinite &&
    north > south && east > west && north <= 90 && south >= -90 && east <= 180 && west >= -180
}

File: mobile/modules/ikkyee-place-search/android/src/main/java/expo/modules/ikkyeeplacesearch/IkkyeePlaceSearchModule.kt
package expo.modules.ikkyeeplacesearch

import android.content.pm.PackageManager
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.RectangularBounds
import com.google.android.libraries.places.api.net.Places
import com.google.android.libraries.places.api.net.PlacesStatusCodes
import com.google.android.libraries.places.api.net.SearchByTextRequest
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class IkkyeePlaceSearchModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("IkkyeePlaceSearch")

    AsyncFunction("searchPlaces") {
      query: String,
      north: Double,
      south: Double,
      east: Double,
      west: Double,
      promise: Promise ->
      val context = appContext.reactContext
      if (context == null || context.packageName != EXPECTED_APPLICATION_ID ||
        !validInput(query, north, south, east, west)) {
        promise.reject("E_PLACE_SEARCH_INVALID", "Native place search input is invalid", null)
        return@AsyncFunction
      }

      val applicationInfo = context.packageManager.getApplicationInfo(
        context.packageName,
        PackageManager.GET_META_DATA
      )
      val apiKey = applicationInfo.metaData?.getString(MAPS_API_KEY_METADATA).orEmpty()
      if (apiKey.isBlank()) {
        promise.reject("E_PLACE_SEARCH_UNAVAILABLE", "Native place search is unavailable", null)
        return@AsyncFunction
      }

      if (!Places.isInitialized()) {
        Places.initializeWithNewPlacesApiEnabled(context.applicationContext, apiKey)
      }
      val fields = listOf(
        Place.Field.ID,
        Place.Field.DISPLAY_NAME,
        Place.Field.FORMATTED_ADDRESS,
        Place.Field.LOCATION
      )
      val request = SearchByTextRequest.builder(query.trim(), fields)
        .setLocationBias(RectangularBounds.newInstance(LatLng(south, west), LatLng(north, east)))
        .setMaxResultCount(MAX_RESULTS)
        .build()

      Places.createClient(context).searchByText(request)
        .addOnSuccessListener { response ->
          val projection = response.places.take(MAX_RESULTS).mapNotNull { place ->
            val id = place.id
            val name = place.displayName
            val location = place.location
            if (id.isNullOrBlank() || name.isNullOrBlank() || location == null) null else mapOf(
              "id" to id,
              "name" to name,
              "address" to (place.formattedAddress?.takeIf { it.isNotBlank() } ?: name),
              "latitude" to location.latitude,
              "longitude" to location.longitude
            )
          }
          promise.resolve(projection)
        }
        .addOnFailureListener { error ->
          val statusCode = (error as? ApiException)?.statusCode
          when (statusCode) {
            PlacesStatusCodes.NOT_FOUND -> promise.resolve(emptyList<Map<String, Any>>())
            PlacesStatusCodes.OVER_QUERY_LIMIT ->
              promise.reject("E_PLACE_SEARCH_QUOTA", "Native place search quota exceeded", null)
            PlacesStatusCodes.REQUEST_DENIED ->
              promise.reject("E_PLACE_SEARCH_CONFIGURATION", "Native place search configuration rejected", null)
            CommonStatusCodes.NETWORK_ERROR ->
              promise.reject("E_PLACE_SEARCH_NETWORK", "Native place search network failed", null)
            else -> promise.reject("E_PLACE_SEARCH", "Native place search failed", null)
          }
        }
    }
  }

  private fun validInput(query: String, north: Double, south: Double, east: Double, west: Double): Boolean {
    return query.trim().isNotEmpty() && query.trim().length <= 80 &&
      listOf(north, south, east, west).all { it.isFinite() } &&
      north > south && east > west && north <= 90 && south >= -90 && east <= 180 && west >= -180
  }

  private companion object {
    const val EXPECTED_APPLICATION_ID = "com.ikkyee.mobile"
    const val MAPS_API_KEY_METADATA = "com.google.android.geo.API_KEY"
    const val MAX_RESULTS = 5
  }
}
